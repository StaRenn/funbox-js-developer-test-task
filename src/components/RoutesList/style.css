.routes-list-wrapper {
  width: 20vw;
  min-width: 280px;
  max-width: 100vw;
  position: absolute;
  max-height: 40vh;
  left: 0;
  top: 0;
  overflow-y: auto;
  z-index: 1;
}

.routes-list {
  padding: 0;
  margin: 0;
  width: 100%;
  word-wrap: break-word;
  height: auto;
}

.routes-list__element__destination {
  width: 75%;
  height: 100%;
  align-items: center;
  color: black;
  display: block;
  word-wrap: break-word;
  border: none;
  outline: none;
  background: none;
  padding: 0 1vh 0 1vh;
  margin: 0;
}

.routes-list__element__delete {
  text-align: center;
  border: none;
  background: none;
  border-left: 2px solid black;
  height: 100%;
  width: 6vw;
  cursor: pointer;
  outline: none;
}

.routes-list__element {
  position: relative;
  box-sizing: border-box;
  background: white;
  width: 100%;
  cursor: grab;
  display: flex;
  align-items: center;
  min-height: 40px;
  height: auto;
  margin-top: -2px;
  border: 2px solid black;
  text-decoration: none;
  list-style-type: none;
  z-index: 1;
}
.routes-list__element * {
  font-family: "Roboto", sans-serif;
  font-size: 1.3rem;
}

/*# sourceMappingURL=style.css.map */
