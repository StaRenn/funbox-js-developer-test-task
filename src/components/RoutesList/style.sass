.routes-list-wrapper
  width: calc(20vw + 4px)
  max-height: 40vh

.routes-list
  padding: 0
  margin: 0

.routes-list__element__destination
  width: calc(20vw - 4vh)
  height: 100%
  color: black
  border: none
  outline: none
  background: none
  padding: 0 1vh 0 1vh
  display: flex
  align-items: center
  margin: 0

.routes-list__element__delete
  text-align: center
  border: none
  background: none
  border-left: 2px solid black
  height: 100%
  width: 6vw
  cursor: pointer
  outline: none

.routes-list__element
  background: white
  width: calc(20vw + 4px)
  cursor: grab
  display: flex
  height: calc(4vh + 4px)
  margin-top: -2px
  border: 2px solid black
  text-decoration: none
  list-style-type: none
  *
    font-family: "Roboto", sans-serif
    font-size: 1.3rem