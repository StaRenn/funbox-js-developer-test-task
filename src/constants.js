export const NEW_ROUTE = "NEW_ROUTE";

export const SET_NEW_ROUTES_ORDER = "SET_NEW_ROUTES_ORDER";

export const DELETE_ROUTE = "DELETE_ROUTE";

export const REPLACE_ROUTE = "REPLACE_ROUTE";

export const YANDEX_MAP_LOADED = "YANDEX_MAPS_LOADED";

export const SET_MAP_CENTER = "SET_MAP_CENTER";