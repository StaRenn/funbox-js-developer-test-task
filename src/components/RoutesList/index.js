import React from 'react';
import "./style.css"
import SearchBar from "../SearchBar";
import {useDispatch, useSelector} from "react-redux";
import {arrayMove, List} from "react-movable";
import {deleteRoute, setNewRoutesOrder} from "../../AC";

function RoutesList() {
    const {routes, order, loading} = useSelector(state => state.routes)
    const orderArr = order.toArray();
    const dispatch = useDispatch();

    const handleDeleteRoute = (id) => () => {
        if(loading) return
        dispatch(deleteRoute(id))
    }

    return (
        <div className={"routes-list-wrapper"}>
            <SearchBar/>
            <List //react-movable component documentation you can see here > github.com/tajo/react-movable
                values={orderArr.map((id) => {
                    return {id: id, name: routes.get(id).name}
                })}
                onChange={({ oldIndex, newIndex }) => {
                    //arrayMove returns new array with new order
                    dispatch(setNewRoutesOrder(arrayMove(orderArr, oldIndex, newIndex)))
                }}
                renderList={({ children, props }) => <ul className={"routes-list"} {...props}>{children}</ul>}
                renderItem={({ value, props }) => {
                    return(
                        <li className={"routes-list__element"} {...props}>
                            <p className={"routes-list__element__destination"}>{value.name}</p>
                            <button className={"action-button"} onClick={handleDeleteRoute(value.id)}>-</button>
                        </li>)}
                }
            />
        </div>
    );
}

export default RoutesList;