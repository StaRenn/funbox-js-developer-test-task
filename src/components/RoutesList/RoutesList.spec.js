import {mount} from "enzyme";
import {Map, List, Record} from "immutable";
import React from "react";
import configureStore from "redux-mock-store"
import {Provider} from "react-redux";
import {DELETE_ROUTE, SET_NEW_ROUTES_ORDER} from "../../constants";
import RoutesList from "./index";

describe('should render and be fully functional ', () => {
    const initialState = {
        routes: Record({
            loading: false,
            center: [55.75, 37.57],
            order: List(["_abc", "_bca"]),
            routes: Map([["_abc", {coords: [55.75, 37,57] ,name: "route1"}], ["_bca", "route2"]])
        })()
    };
    const middlewares = [];
    let mockStore = configureStore(middlewares);
    let store = mockStore(initialState);
    let component;
    const expectedActions = {
        deleteRoute: [{
            type: DELETE_ROUTE,
            payload: {id: "_bca"}
        }],
        setNewRoutesOrder: [{
            type: SET_NEW_ROUTES_ORDER,
            payload: ["_bca", "_abc"]
        }]
    }

    beforeEach(() => {
        component = mount(<Provider store={store}><RoutesList /></Provider>);
        store.clearActions();
    });

    it("should render and match snapshot", () => {
        expect(component).toMatchSnapshot();
    });

    it("should render 2 routes", () => {
        expect(component.find(".routes-list__element").length).toBe(2);
    })

    it("should dispatch DELETE_ROUTE on delete button click", () => {
        component.find(".routes-list__element").at(1).find("button").simulate("click");
        expect(store.getActions()).toEqual(expectedActions.deleteRoute);
    })
})