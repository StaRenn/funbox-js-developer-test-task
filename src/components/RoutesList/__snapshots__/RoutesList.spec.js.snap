// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render and be fully functional  should render and match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <RoutesList>
    <div
      className="routes-list-wrapper"
    >
      <SearchBar>
        <div
          className="search-bar"
        >
          <input
            className="search-bar__input"
            onChange={[Function]}
            onKeyDown={[Function]}
            value=""
          />
          <button
            className="action-button"
            onClick={[Function]}
          >
            +
          </button>
        </div>
      </SearchBar>
      <List
        lockVertically={false}
        onChange={[Function]}
        removableByMove={false}
        renderItem={[Function]}
        renderList={[Function]}
        transitionDuration={300}
        values={
          Array [
            Object {
              "id": "_abc",
              "name": "route1",
            },
            Object {
              "id": "_bca",
              "name": undefined,
            },
          ]
        }
        voiceover={
          Object {
            "canceled": [Function],
            "dropped": [Function],
            "item": [Function],
            "lifted": [Function],
            "moved": [Function],
          }
        }
      >
        <ul
          className="routes-list"
        >
          <li
            aria-roledescription="You are currently at a draggable item at position 1. Press space bar to lift."
            className="routes-list__element"
            key="0"
            onKeyDown={[Function]}
            style={
              Object {
                "MozUserSelect": "none",
                "WebkitUserSelect": "none",
                "boxSizing": "border-box",
                "msUserSelect": "none",
                "position": "relative",
                "touchAction": "none",
                "userSelect": "none",
                "visibility": undefined,
                "zIndex": 0,
              }
            }
            tabIndex={0}
          >
            <p
              className="routes-list__element__destination"
            >
              route1
            </p>
            <button
              className="action-button"
              onClick={[Function]}
            >
              -
            </button>
          </li>
          <li
            aria-roledescription="You are currently at a draggable item at position 2. Press space bar to lift."
            className="routes-list__element"
            key="1"
            onKeyDown={[Function]}
            style={
              Object {
                "MozUserSelect": "none",
                "WebkitUserSelect": "none",
                "boxSizing": "border-box",
                "msUserSelect": "none",
                "position": "relative",
                "touchAction": "none",
                "userSelect": "none",
                "visibility": undefined,
                "zIndex": 0,
              }
            }
            tabIndex={0}
          >
            <p
              className="routes-list__element__destination"
            />
            <button
              className="action-button"
              onClick={[Function]}
            >
              -
            </button>
          </li>
        </ul>
        <div
          aria-atomic="true"
          aria-live="assertive"
          role="log"
          style={
            Object {
              "border": "0px",
              "clip": "react(0px, 0px, 0px, 0px)",
              "clipPath": "inset(100%)",
              "height": "1px",
              "margin": "-1px",
              "overflow": "hidden",
              "padding": "0px",
              "position": "absolute",
              "width": "1px",
            }
          }
        />
      </List>
    </div>
  </RoutesList>
</Provider>
`;
